File: programmers/lv3/정수 삼각형.cpp
//https://school.programmers.co.kr/learn/courses/30/lessons/43105?language=cpp
//정수 삼각형
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> triangle) {
    //시도 1. Top to Bottom
    //시도 2. Bottom to Top
    vector<vector<int>> sum = triangle;
    for(int h=sum.size()-2; h>=0; h--){
        for(int w=0; w<sum[h].size(); w++){
            int left = sum[h+1][w];
            int right = sum[h+1][w+1];
            if(left>right){
                sum[h][w] = sum[h][w]+left;
            }else{
                sum[h][w] = sum[h][w]+right;
            }
        }
    }
    return sum[0][0];
}

File: programmers/lv3/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int depth = triangle.length-1;
        
        for(int d = depth; d>0; d--){
            for(int n = 0; n<triangle[d].length-1; n++){
                int pn = triangle[d-1][n];
                triangle[d-1][n] = Math.max(triangle[d][n]+pn,triangle[d][n+1]+pn);
            }
        }
        return triangle[0][0];
    }
}
